File: example/websockets/wsserver.cc
//
//  wsserver.cc
//
//  Copyright (c) 2025 Florian Albrechtskirchinger. All rights reserved.
//  MIT License
//

#include "httplib-ws.h"

int main(void) {
  httplib::WebSocketServer svr;

  svr.WebSocket("/endpoint", [](httplib::WSConnection &connection, std::string msg, httplib::WSMessageType msg_type) {
    std::cout << "Message received: " << msg << "\n";
    connection.wait_until_ready();
    if(connection.send(std::move(msg), msg_type)) {
      std::cout << "Sucessfully echoed message!\n";
    }
  }, httplib::WSSubprotocols{"echo"});

  std::cout << "Listening on http://localhost:8080/\n";

  svr.listen("localhost", 8080);

  return 0;
}

File: example/websockets/wsserver.py
#!/usr/bin/env python3
#
#  wsclient.py
#
#  Copyright (c) 2025 Florian Albrechtskirchinger. All rights reserved.
#  MIT License
#

import asyncio
import logging
import websockets


async def echo(websocket):
    try:
        async for message in websocket:
            # Echo each message back to the client
            await websocket.send(message)
            print(f"Received and echoed: {message}")
    except websockets.exceptions.ConnectionClosedOK:
        print("Client disconnected normally")
    except Exception as e:
        print(f"Error in connection: {e}")


async def main():
    # Start the server on localhost at port 8080
    server = await websockets.serve(echo, "localhost", 8080, ping_interval=2)
    print("WebSocket Echo Server started on ws://localhost:8080")
    await server.wait_closed()


if __name__ == "__main__":
    logging.basicConfig(
        format="%(message)s",
        level=logging.DEBUG,
    )
    asyncio.run(main())
